tinytga/src/lib.rs

```diff
@@ -93,324 +110,216 @@
+        let (input, header) = TgaHeader::parse(self.data).ok()?;
+
+        image_id(input, &header)
+            .ok()
+            .map(|(_input, id)| id)
+            .filter(|id| !id.is_empty())
+    }
+}
 
-    /// Number of bytes contained within each pixel
-    stride: usize,
+impl<'a, C> Tga<'a, C>
+where
+    C: PixelColor,
+{
+    /// Parse a TGA image from a byte slice
+    pub fn from_slice(data: &'a [u8]) -> Result<Self, ParseError> {
+        let tga = Tga::from_slice_common(data)?;
 
-    /// Current X coordinate from top-left of image
-    x: u16,
+        if C::Raw::BITS_PER_PIXEL != usize::from(tga.color_bpp().bits()) {
+            return Err(ParseError::MismatchedBpp(tga.color_bpp().bits()));
+        }
 
-    /// Current Y coordinate from top-left of image
-    y: u16,
+        Ok(tga)
+    }
 
-    /// Iteration is done
-    done: bool,
+    /// Returns an iterator over the raw pixels in this image.
+    pub fn pixels<'b>(&'b self) -> Pixels<'b, 'a, C> {
+        Pixels::new(self.raw_pixels())
+    }
 }
 
-impl<'a> Iterator for TgaIterator<'a> {
-    type Item = Pixel;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        if self.done {
-            return None;
-        }
-
-        if self.current_packet_position >= self.current_packet_pixel_length {
-            // Parse next packet from remaining bytes
-            match self.tga.header.image_type {
-                ImageType::Monochrome | ImageType::Truecolor | ImageType::ColorMapped => {
-                    return None;
-                }
-                ImageType::RleMonochrome | ImageType::RleTruecolor | ImageType::RleColorMapped => {
-                    if self.bytes_to_consume.is_none() {
-                        return None;
-                    } else {
-                        self.current_packet_position = 0;
-                    }
-
-                    let (bytes_to_consume, current_packet) =
-                        next_rle_packet(self.bytes_to_consume.unwrap(), self.tga.bpp() / 8)
-                            .map(|(remaining, packet)| {
-                                (
-                                    if !remaining.is_empty() {
-                                        Some(remaining)
-                                    } else {
-                                        None
-                                    },
-                                    packet,
-                                )
-                            })
-                            .expect("Failed to parse first image RLE data packet");
-
-                    self.bytes_to_consume = bytes_to_consume;
-                    self.current_packet_pixel_length = current_packet.len() / self.stride;
-                    self.current_packet = current_packet;
-                }
-                image_type => panic!("Image type {:?} not supported", image_type),
-            };
-            // }
-        }
+impl<'a> Tga<'a, ()> {
+    /// Parse a TGA image from a byte slice
```

Comment:
This could use some clarification around colour depth validation and/or its differences from `from_slice`.

Reply:
As I said in the original comment the docs still need some work. Would it be OK to add this in the next PR? The additional `DynamicTga` type will make this even more confusing and there should be one overview to show when to use which type IMO.

Reply:
Yeah that's fine. I added some comment suggestions in case you wanted to add some of them in now.

Reply:
Thanks for the suggestions, I've already added them. I didn't mean that these changes weren't appreciated, but I didn't want to cause extra work if I decide to do bigger changes in the next PR which could make some of the added docs invalid.